Great refactor of the build definition

odisseus opened the pull request:
The old build had a number of problems:
 * NPM dependencies were defined in two different places. As a consequence, the versions used by the Scala.js compiler didn't match the versions required by the extension manifest.
 * The build artifacts were being placed outside the `target` directory, and `sbt clean open` didn't actually clean them. This could cause a lot of mistakes and confusion when testing new changes.
 * The build redefined parts of the functionality provided by existing SBT plugins.

The new build has these problems fixed:
 * Every configuration value and dependency version is defined in a single place.
 * All of the above are visible to the SBT build.
 * When you clean the build artifacts, they actually get cleaned.
 * Instead of defining a custom task, the build uses `npmInstallDependencies` from the sbt-scalajs-bundler plugin.

odisseus committed:
Great refactor of the build definition

The old build had a number of problems:
 * NPM dependencies were defined in two different places. As a consequence, the versions used by the Scala.js compiler didn't match the versions required by the extension manifest.
 * The build artifacts were being placed outside the `target` directory, and `sbt clean open` didn't actually clean them. This could cause a lot of mistakes and confusion when testing new changes.
 * The build redefined parts of the functionality provided by existing SBT plugins.

The new build has these problems fixed:
 * Every configuration value and dependency version is defined in a single place.
 * All of the above are visible to the SBT build.
 * When you clean the build artifacts, they actually get cleaned.
 * Instead of defining a custom task, the build uses `npmInstallDependencies` from the sbt-scalajs-bundler plugin.

Baccata commented:
Hey, thanks for this ! I was away the past few days, thanks for your patience. I'll give it a try today. 

While we're talking : do you mind if I ask about the context of this work ? Don't get me wrong, I appreciate the contribution, but I'm not sure where the motivation comes from, considering this is housecleaning work. Are you interested in adding functionality later ?

odisseus commented:
I am writing my own VS Code extension (which isn't published yet), and I have found that there is no standard way to do it with Scala.js and SBT. I have used your project and [@pme123/vscode-scalajs-hello](https://github.com/pme123/vscode-scalajs-hello) as examples, but I needed both the JS part and the Scala part (client and server) configured in the same multi-project build.

I have found out that the easiest way to achieve this is to write a SBT plugin that would configure the extension project and define the additional tasks. Having applied this solution to my own project, I decided to share it with other projects that could make use of it.

I consider publishing said plugin as a Maven artifact at some point. However, my own extension project is at a very early stage, and it doesn't yet use many of the features that are used by other projects such as yours. Thus I am also interested in getting my code tested and obtaining the feedback.

One feature that I'm going to add later is publishing to the VS Code Marketplace. Looks like [@scalacenter/accessible-scala](https://github.com/scalacenter/accessible-scala) already has it implemented, so I think I could reuse their code.

Baccata commented:
Awesome, thanks for the context. My day has been derailed a bit and unfortunately I've not had time to give it a try today, but I'll definitely do so over the weekend !

odisseus commented:
Sorry for taking your time, but I'm really interested in merging this, so that I may better understand any possible problems with my approach. I will be happy to answer any questions you might have.

Baccata commented:
Hi there, sorry I haven't replied until now, things have been really chaotic at work (I work for Disney) and I admit I had forgotten to send you feedback because of it :( .

Anyhow, after a little bit of thought : I'm not certain that's the direction I'd like to take : having the data that is usually held by `package.json` in SBT helps unify the source of truth, but there's a certain amount of functionality that is lost, such as github's dependabot integration / `npm update`. I know that some other solutions tend to use `package.json` as a source of truth and source settings from there .. 

I don't think it's necessarily a dealbreaker though, it's a very low-traffic project, and I do appreciate the contribution. I just want to be able to test it, but scaladex has been a little bit flaky over the past week and seems down at the moment, which doesn't help the testing 😅.

Baccata commented:
Alright I've tested and mostly it works fine, and I'm reasonably happy with the workflow it provides, but I can't merge the PR until `vsce package` and `vsce publish` are handled by SBT somehow

odisseus commented:
The complete build output is assembled in the directory `target/scala-2.13/scalajs-bundler/main`. (I have added a new command for doing that.) When I run `vsce package` in that directory, an installable VSIX file is generated. I haven't tested `vsce publish` (because, for one thing, I haven't registered as a publisher), but I assume it would work as well.

As for Dependabot, I'm afraid it is plainly impossible to use it when the `package.json` file is dynamically generated. There has been a [suggestion](https://github.com/scalacenter/scalajs-bundler/issues/310) that `package.json` should be supported by the Scala.js bundler plugin specifically for this reason, but apparently it didn't gain traction.

Baccata commented:
> The complete build output is assembled in the directory target/scala-2.13/scalajs-bundler/main. (I have added a new command for doing that.) When I run vsce package in that directory, an installable VSIX file is generated. I haven't tested vsce publish (because, for one thing, I haven't registered as a publisher), but I assume it would work as well.

I've given this a go and the resulting vsix file is around 10MB large, whilst the workflow on the main branch gives me a ~160KB large file. I presume something is off (possibly due to typescript appearing as a dependency in the package.json when it's not the case). 

I cannot accept the PR without this being addressed ... : even if the workflow seems nicer in surface, the resulting artifact should be of the same size as what the current workflow gives me. 

Additionally, my point wrt `vsce package` and `vsce publish` commands is that without bespoke SBT commands handling that for me, the workflow provided by an SBT plugin is incomplete imho : I don't want to have to `cd` to the correct directory in order to package the vsix file.  However, I'm willing to overlook the absence of these bespoke commands in this PR and implement them myself.

Baccata commented:
On a totally unrelated note : it's usually deemed a bad idea to amend commits / force-push after you've opened a pull-request and people started commenting on it. It makes it really hard for the reviewer to understand what changes in between iterations and whether their comments are addressed at all. 

It's much preferable to just send normal commits, and ask the blessing of the maintainer/reviewer to squash the commits before the PR is merged.

odisseus commented:
I found out that Typescript is being added due to a problem in the sbt-scalajs-bundler plugin. I have created an [issue](https://github.com/ScalablyTyped/Converter/issues/524) for that.

odisseus commented:
The latest beta of `org.scalablytyped.converter`, which has been [released](https://github.com/ScalablyTyped/Converter/releases/tag/v1.0.0-beta43) a couple weeks ago, fixes the problem of excessive VSIX size (now it is about 260 KiB large). I have updated my changes accordingly and rebased them on top of the latest main. Can you please review this PR again?

kubukoz commented on project/VsCodeExtensionPlugin.scala:
```diff
@@ -0,0 +1,80 @@
+
+import scala.sys.process._
+
+object VsCodeExtensionPlugin extends AutoPlugin {
```
praise: this is really nice!

kubukoz reviewed (commented):
bumping this one - I was going to try and bump some deps (and maybe go scala 3), but this would be nice to get in first.

Baccata commented:
I'm gonna be honest : after giving it a think, I'm just not particularly interested in merging this, as I don't think managing npm dependencies in SBT is a worthwhile endeavour here. 

The current workflow is satisfying to me, considering how infrequently I need to touch this project. Going forward, I think it'd be better serving to the community if this project was folded directly in metals, but I have neither the time nor drive to take that on. 

I apologise for not getting back to you earlier. I appreciate the intent, but the reality of open-source is that maintainers don't have to accommodate the vision of perfection of everyone. If you have a vision for what this project should be, feel free to fork, you have my blessing.
